snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+
+            # TODO: This timeout was a totally arbitrary descision and may or
+            # may not make sense.
+            # TODO: This may need to be protected from EndOfStream errors,
+            # depending on how that configuration has been set.
+            message = self.__commit_log_consumer_backend.poll(0.1)
+            if message is None:
+                continue
+
+            # If this update matches our subscription (whether or not it is
+            # part of the current assignment does not matter -- it may be part
+            # of a later assignment) and was from a remote consumer group that
+            # we are observing, we need to update the offset state. First, we
+            # check the consumer groups reference (because we can do so without
+            # taking the lock):
+            if message.value.consumer_group in self.__remote_consumer_groups:
+                with self.__lock:
+                    # If the current subscription has changed while we were
+                    # waiting for a message, we need to discard this message.
+                    # The next loop iteration will update the consumer
+                    # assignment, and we will begin to rebuild state there.
+                    if self.__subscription is not current_subscription:
+                        continue
+
+                    # If the stream for this message matches the subscription
+                    # topics, apply the update to the subscription offsets.
+                    if any(
+                        (message.stream in topic)
+                        for topic in self.__subscription.topics
+                    ):
+                        self.__subscription.remote_consumer_group_offsets[
+                            message.value.stream
+                        ][message.value.consumer_group] = message.value.offset
+
+    def __get_effective_remote_offset(self, stream: TStream) -> Optional[TOffset]:
+        # XXX: This assumes that the lock has already been acquired before
+        # calling this method.
+        offsets = self.__subscription.remote_consumer_group_offsets[stream]
+        return min(
+            filter(
+                lambda offset: offset is not None,
+                [offsets.get(group) for group in self.__remote_consumer_groups],
+            ),
+            default=None,
+        )
+
+    def subscribe(
+        self,
+        topics: Sequence[
+            Topic[TStream]
+        ],  # TODO: This will have to apply to the superclass as well.
+        on_assign: Optional[Callable[[Mapping[TStream, TOffset]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        # Update the subscription, so that the commit log consumer can starts
+        # to consume the commit log data for the subscribed (and potentially
+        # assigned) topics.
+        with self.__lock:
+            self.__subscription = Subscription(topics)
```

Comment:
It took me a while to realize this: it's drastically simpler for the commit log consumers to just subscribe to all offsets that *may* be assigned to this consumer while the current subscription is active, rather than clearing out and re-initializing offsets for only the owned partition after every rebalance.